File: sendfile-Attachments/workflows-java/src/main/java/net/corda/examples/sendfile/flows/DownloadAttachment.java
package net.corda.examples.sendfile.flows;

import co.paralleluniverse.fibers.Suspendable;
import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.Attachment;
import net.corda.core.crypto.SecureHash;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowLogic;
import net.corda.core.flows.InitiatingFlow;
import net.corda.core.flows.StartableByRPC;
import net.corda.core.identity.Party;
import net.corda.core.node.services.Vault;
import net.corda.core.node.services.vault.QueryCriteria;
import net.corda.core.utilities.ProgressTracker;
import net.corda.examples.sendfile.states.InvoiceState;
import org.jetbrains.annotations.Nullable;

import java.io.*;

// *********
// * Flows *
// *********
@InitiatingFlow
@StartableByRPC
public class DownloadAttachment extends FlowLogic<String> {
    private final Party sender;
    private final String path;

    private final ProgressTracker.Step RETRIEVING_ID = new ProgressTracker.Step("Retrieving attachment ID");
    private final ProgressTracker.Step DOWNLOAD_ATTACHMENT = new ProgressTracker.Step("Download attachment");

    private final ProgressTracker progressTracker = new ProgressTracker(RETRIEVING_ID, DOWNLOAD_ATTACHMENT);

    @Nullable
    @Override
    public ProgressTracker getProgressTracker() {
        return progressTracker;
    }

    public DownloadAttachment(Party sender, String path) {
        this.sender = sender;
        this.path = path;
    }

    @Suspendable
    @Override
    public String call() throws FlowException {
        progressTracker.setCurrentStep(RETRIEVING_ID);
        QueryCriteria criteria = new QueryCriteria.VaultQueryCriteria()
                .withParticipants(ImmutableList.of(sender, getOurIdentity()));

        String state = getServiceHub().getVaultService().queryBy(
                InvoiceState.class,
                criteria
        ).getStates().get(0).getState().getData().getInvoiceAttachmentID();

        progressTracker.setCurrentStep(DOWNLOAD_ATTACHMENT);
        Attachment content = getServiceHub().getAttachments().openAttachment(SecureHash.parse(state));
        try {
            assert content != null;
            //content.extractFile(path, new FileOutputStream(new File(path)));
            InputStream inStream = content.open();
            byte[] buffer = new byte[inStream.available()];
            inStream.read(buffer);
            File targetFile = new File(path);
            new FileOutputStream(targetFile).write(buffer);
        } catch (IOException e) {
            e.printStackTrace();
        }

        return "Downloaded file from " + sender.getName().getOrganisation() + " to " + path;
    }
}

File: sendfile-Attachments/workflows-kotlin/src/main/kotlin/net/corda/examples/sendfile/flows/DownloadAttachment.kt
package net.corda.examples.sendfile.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.examples.sendfile.states.InvoiceState
import net.corda.core.crypto.SecureHash
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.node.services.vault.QueryCriteria
import net.corda.core.utilities.ProgressTracker
import java.io.File
import java.io.InputStream

// *********
// * Flows *
// *********
@InitiatingFlow
@StartableByRPC
class DownloadAttachment(
        private val sender: Party,
        private val path: String
) : FlowLogic<String>() {
    companion object {
        object RETRIEVING_ID : ProgressTracker.Step("Retrieving attachment ID")
        object DOWNLOAD_ATTACHMENT : ProgressTracker.Step("Download attachment")

        fun tracker() = ProgressTracker(
                RETRIEVING_ID,
                DOWNLOAD_ATTACHMENT
        )
    }

    override val progressTracker = tracker()
    @Suspendable
    override fun call():String {
        progressTracker.currentStep = RETRIEVING_ID
        val criteria = QueryCriteria.VaultQueryCriteria(
                participants = listOf(sender,ourIdentity)
        )

        val state = serviceHub.vaultService.queryBy(
                contractStateType = InvoiceState::class.java,
                criteria = criteria
        ).states.get(0).state.data.invoiceAttachmentID

        progressTracker.currentStep = DOWNLOAD_ATTACHMENT
        val content = serviceHub.attachments.openAttachment(SecureHash.parse(state))!!
        content.open().toFile(path)

        return "Downloaded file from " + sender.name.organisation + " to " + path
    }
}


fun InputStream.toFile(path: String) {
    File(path).outputStream().use { this.copyTo(it) }
}

